add authentication docs page (#60388)
vercel/next.js: docs/02-app/01-building-your-application/09-authentication/index.mdx

Here's how to implement middleware for authentication in Next.js:

1. **Setting Up Middleware:**
   - Create a `middleware.ts` or `.js` file in your project's root directory.
   - Include logic to authorize user access, such as checking for authentication tokens.
2. **Defining Protected Routes:**
   - Use the `matcher` option in your middleware to identify routes that require authorization checks.
3. **Middleware Logic:**
   - Write logic to verify if a user is authenticated. Check user roles or permissions for route authorization.
4. **Handling Unauthorized Access:**
   - Redirect unauthorized users to a login or error page as appropriate.

Example middleware file:

```ts filename="middleware.ts" switcher
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  const currentUser = request.cookies.get('currentUser')?.value

  if (currentUser) {
    return NextResponse.redirect(new URL('/dashboard', request.url))
  } else {
    return NextResponse.redirect(new URL('/login', request.url))
  }
}

export const config = {

>   matcher: '/dashboard',

Reviewer: Maybe we can use the matcher from `/learn` or the one recommended by Clerk? https://clerk.com/docs/references/nextjs/auth-middleware#usage

This example uses [`Response.redirect`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirect_static) for efficient, centralized access control. It handles redirects early in the request pipeline, making it efficient and ideal for broad strategies.

For specific redirection needs, the `redirect` function can be used in Server Components, Route Handlers, and Server Actions to provide more control. This is useful for role-based navigation or context-sensitive scenarios.

```js
import { redirect } from 'next/navigation'

export default function Page() {
  // Logic to determine if a redirect is needed
  const accessDenied = true
  if (accessDenied) {
    redirect('/login')
  }

  // Define other routes and logic
}
```

> This flexibility is key when handling redirections based on detailed authentication outcomes or other contextual factors. For example, `redirect('/login')` in a Server Action allows for precise redirection based on the user's role or specific conditions during the authentication process.

Reviewer: Sorry, I'm not sure I understand this part. What are detailed authentication outcomes and contextual factors? Could we possibly provide use cases?

### Protecting Server Actions

It is important to treat [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) with the same security considerations as public-facing API endpoints. Verifying user authorization for each action is crucial. Implement checks within Server Actions to determine user permissions, such as restricting certain actions to admin users.

In the example below, we check the user's role before allowing the action to proceed:

> ```ts filename="app/lib/actions.ts" switcher

Reviewer: Can we add the JS equivalent?

'use server'

// ...

> export async function serverAction(request) {

Reviewer: - The `request` argument is not being used in the function, you can delete it. 
- Consider calling the action something that reflects a real-world example instead of `serverAction`.

```suggestion
export async function serverAction() {
```

  const session = await getSession()
  const userRole = session?.user?.role

  // Check if user is authorized to perform the action
  if (userRole !== 'admin') {
    throw new Error('Unauthorized access: User does not have admin privileges.')
  }

  // Proceed with the action for authorized users
  // ... implementation of the action
}
```

### Protecting Route Handlers

Route Handlers in Next.js play a vital role in managing incoming requests. Just like server actions, they should be secured to ensure that only authorized users can access certain functionalities. This often involves verifying the user's authentication status and their permissions.

Here's an example of securing a Route Handler:

```ts filename="app/api/route.ts" switcher
export async function GET() {
  // User authentication and role verification
  const session = await getSession()

  // Check if the user is authenticated
  if (!session) {
    return new Response(null, { status: 401 }) // User is not authenticated
  }

  // Check if the user has the 'admin' role
  if (session.user.role !== 'admin') {

>     return new Response(null, { status: 403 }) // User is authenticated but does not have the right permissions

Reviewer: Nice status code 👍🏼

  }

  // Data fetching for authorized users

>   // {...}

Reviewer: ```suggestion
```

}
```

This example demonstrates a Route Handler with a two-tier security check for authentication and authorization. It first checks for an active session, and then verifies if the logged-in user is an 'admin'. This approach ensures secure access, limited to authenticated and authorized users, maintaining robust security for request processing.

### Authorization Using Server Components

> [Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) in Next.js are designed for server-side execution and offer a secure environment for integrating complex logic like authorization. They enable direct access to back-end resources, optimizing performance for data-heavy tasks and enhancing security for sensitive operations.

Reviewer: ```suggestion
[Server Components](docs/app/building-your-application/rendering/server-components) in Next.js are designed for server-side execution and offer a secure environment for integrating complex logic like authorization. They enable direct access to back-end resources, optimizing performance for data-heavy tasks and enhancing security for sensitive operations.
```

**Key Advantages of Server Components:**

- Server-Side Execution: Offloads complex tasks from the client, improving performance.
- Direct Back-End Access: Simplifies interactions with server resources, reducing the need for extra API endpoints.

> - Enhanced Security: Ideal for critical operations like authorization, as server-side execution minimizes exposure to vulnerabilities.

Reviewer: You could remove _the advantages using Server Components_ as it's a tangent to auth. It should be ok as the benefits are outlined in the Server Components docs you linked above.

**Implementing Authorization Logic in Server Components**

Implementing authorization in Server Components involves verifying user roles or permissions and conditionally rendering content based on these verifications.

1. **Identify Authorization Requirements**: Determine restricted access areas in your application.
2. **Fetch User Details**: Retrieve user roles or permissions from sessions or authentication contexts.
3. **Role-Based Access Control**: Check user roles or permissions using server components.
4. **Conditional Rendering**: Render components based on user authorization.

> 5. **Handle Unauthorized Access**: Display messages or redirect for insufficient permissions.

Reviewer: I think you can also remove this section as it's not very actionable. 

Based on the example, perhaps what should be highlighted here is that, in Server Component, you can conditionally render UI based on the user role?

**Example:**

```tsx filename="app/dashboard/page.tsx" switcher

> // app/dashboard/page.ts

Reviewer: Can remove as filename is above.

```suggestion
```

> const Dashboard = async () => {

Reviewer: Prefer functions, should be a `default export` in page.tsx.

```suggestion
export default function Dashboard() => {
```

### Cookie-Based Sessions

Cookie-based sessions manage user data by storing encrypted session information directly in browser cookies. Upon user login, this encrypted data is stored in the cookie. Each subsequent server request includes this cookie, minimizing the need for repeated server queries and enhancing client-side efficiency. However, this method requires careful encryption to protect sensitive data, as cookies are susceptible to client-side security risks. Additionally, while individual cookies are limited in size (typically around 4KB), techniques like cookie-chunking can overcome this limitation by dividing large session data into multiple cookies.

Setting a cookie in a Next.js project might look something like this:

**Setting a cookie on the server:**

```jsx

> import { setCookie } from 'next/cookies'

Reviewer 2: ```suggestion
import { cookies } from 'next/headers'
```

> export function handleLogin(req, res, sessionData) {

Reviewer 2: `req` and `res` are not being used here

  const encryptedSessionData = encrypt(sessionData) // Encrypt your session data

>   setCookie({ res }, 'session', encryptedSessionData, {

Reviewer 2: ```suggestion
  cookies().set('session', encryptedSessionData, {
```

https://nextjs.org/docs/app/api-reference/functions/cookies#cookiessetname-value-options

    httpOnly: true,
    secure: process.env.NODE_ENV === 'production',
    maxAge: 60 * 60 * 24 * 7, // One week
    path: '/',
  })
  // Redirect or handle the response after setting the cookie
}
```

**Accessing the session data stored in the cookie in a server component:**

```jsx

> import { getCookie } from 'next/cookies'

Reviewer 2: ```suggestion
import { cookies } from 'next/headers'
```

export async function getSessionData(req) {

>   const encryptedSessionData = getCookie(req, 'session')

Reviewer 2: ```suggestion
  const encryptedSessionData = cookies().get('session')?.value
```

https://nextjs.org/docs/app/api-reference/functions/cookies#cookiesgetname

However, this approach has its trade-offs. It can increase performance overhead due to the need for database lookups at each user interaction. Strategies like session data caching can help mitigate this. Additionally, reliance on the database means that session management is as reliable as the database’s performance and availability.

Here's a simplified example of implementing database sessions in a Next.js application:

**Creating a Session on the Server**:

```js
import db from 'path/to/databaseModule'

export async function createSession(user) {
  const sessionId = generateSessionId() // Generate a unique session ID
  await db.insertSession({ sessionId, userId: user.id, createdAt: new Date() })
  return sessionId
}
```

**Retrieving a Session in Middleware or Server-Side Logic**:

```js

> import db from 'path/to/databaseModule'
> 
> export async function getSession(req) {
>   const sessionId = req.cookies.sessionId // Assuming session ID is stored in a cookie

Reviewer 2: ```suggestion
import { cookies } from 'next/headers'
import db from 'path/to/databaseModule'

export async function getSession() {
  const sessionId = cookies().get("sessionId")?.value // Assuming session ID is stored in a cookie
```

  return sessionId ? await db.findSession(sessionId) : null
}
```

### Selecting Session Management in Next.js

Deciding between cookie-based and database sessions in Next.js depends on your application's needs. Cookie-based sessions are simpler and suit smaller applications with lower server load but may offer less security. Database sessions, while more complex, provide better security and scalability, ideal for larger, data-sensitive applications.

> With [authentication solutions](https://www.notion.so/Authentication-fec5315fd5de47988e45d0ee1996e3c5?pvs=21) such as NextAuth.js, session management becomes more efficient, using either cookies or database storage. This automation simplifies the development process, but it's important to understand the session management method used by your chosen solution. Ensure it aligns with your application's security and performance requirements.

Reviewer 2: ```suggestion
With [authentication solutions](https://www.notion.so/Authentication-fec5315fd5de47988e45d0ee1996e3c5?pvs=21) such as NextAuth.js, session management becomes more efficient, using either cookies or database storage. This automation simplifies the development process, but it's important to understand the session management method used by your chosen solution. Ensure it aligns with your application's security and performance requirements.
```

Link should be changed to a public URL

Regardless of your choice, prioritize security in your session management strategy. For cookie-based sessions, using secure and HTTP-only cookies is crucial to protect session data. For database sessions, regular backups and secure handling of session data are essential. Implementing session expiry and cleanup mechanisms is vital in both approaches to prevent unauthorized access and maintain application performance and reliability.

## Authentication Solutions

Here are authentication solutions compatible with Next.js, please refer to the quickstart guides below to learn how to configure them in your Next.js application:

> - [NextAuth.js](https://next-auth.js.org/configuration/initialization#route-handlers-app)
> - [AuthO](https://auth0.com/docs/quickstart/webapp/nextjs/01-login)
> - [Supabase](https://supabase.com/docs/guides/getting-started/quickstarts/nextjs)
> - [Stytch](https://stytch.com/docs/guides/quickstarts/nextjs)
> - [Clerk](https://clerk.com/docs/quickstarts/nextjs)

Reviewer 2: ```suggestion
- [AuthO](https://auth0.com/docs/quickstart/webapp/nextjs/01-login)
- [Clerk](https://clerk.com/docs/quickstarts/nextjs)
  {/* TODO: Change link to authjs.dev when new documentation is ready */}
- [NextAuth.js](https://next-auth.js.org/configuration/initialization#route-handlers-app)
- [Supabase](https://supabase.com/docs/guides/getting-started/quickstarts/nextjs)
- [Stytch](https://stytch.com/docs/guides/quickstarts/nextjs)
```

Sorted alphabetically

For specific redirection needs, the `redirect` function can be used in Server Components, Route Handlers, and Server Actions to provide more control. This is useful for role-based navigation or context-sensitive scenarios.

```ts filename="/app/page.tsx" switcher
import { redirect } from 'next/navigation'

export default function Page() {
  // Logic to determine if a redirect is needed
  const accessDenied = true
  if (accessDenied) {
    redirect('/login')
  }

  // Define other routes and logic
}
```

```js filename="/app/page.jsx" switcher
import { redirect } from 'next/navigation'

export default function Page() {
  // Logic to determine if a redirect is needed
  const accessDenied = true
  if (accessDenied) {
    redirect('/login')
  }

  // Define other routes and logic
}
```

This flexibility is crucial for role-specific redirections post-authentication. For example, an admin user can be redirected to an admin dashboard, while a regular user might be sent to a different page. It's also useful for conditional navigation, like redirecting users to complete their profile if necessary.

> When setting up authorization, it's important to make sure the main security checks happen where your app accesses or changes data. Middleware can help by doing a first check, but it shouldn't be the only thing keeping your data safe. Always check if a user has the right access when you're actually getting or changing the data. This way, even if someone gets past the first check, your data is still secure at the most important step - when it's being accessed.

Author: I added this section based on the conve we had with @leerob regarding https://vercel.slack.com/archives/C03KAR5DCKC/p1704943720619029?thread_ts=1704942838.625119&cid=C03KAR5DCKC. What do you think? I kept it brief, is it worth going into more detail?

### Cookie-Based Sessions

Cookie-based sessions manage user data by storing encrypted session information directly in browser cookies. Upon user login, this encrypted data is stored in the cookie. Each subsequent server request includes this cookie, minimizing the need for repeated server queries and enhancing client-side efficiency. However, this method requires careful encryption to protect sensitive data, as cookies are susceptible to client-side security risks. Additionally, while individual cookies are limited in size (typically around 4KB), techniques like cookie-chunking can overcome this limitation by dividing large session data into multiple cookies.

Setting a cookie in a Next.js project might look something like this:

**Setting a cookie on the server:**

```tsx filename="/app/page.tsx" switcher
import { cookies } from 'next/headers'

export async function handleLogin(sessionData) {

>   const encryptedSessionData = encrypt(sessionData) // Encrypt your session data

Reviewer: Do we want to explain why we're encrypting session data? Sorry, I'm not sure I understand it.

  cookies().set('session', encryptedSessionData, {
    httpOnly: true,
    secure: process.env.NODE_ENV === 'production',
    maxAge: 60 * 60 * 24 * 7, // One week
    path: '/',
  })
  // Redirect or handle the response after setting the cookie
}
```

```jsx filename="/app/page.jsx" switcher
import { cookies } from 'next/headers'

export async function handleLogin(sessionData) {
  const encryptedSessionData = encrypt(sessionData) // Encrypt your session data
  cookies().set('session', encryptedSessionData, {
    httpOnly: true,
    secure: process.env.NODE_ENV === 'production',
    maxAge: 60 * 60 * 24 * 7, // One week
    path: '/',
  })
  // Redirect or handle the response after setting the cookie
}
```

**Accessing the session data stored in the cookie in a server component:**

```tsx filename="/app/page.tsx" switcher
import { cookies } from 'next/headers'

export async function getSessionData(req) {
  const encryptedSessionData = cookies().get('session')?.value
  return encryptedSessionData ? JSON.parse(decrypt(encryptedSessionData)) : null
}

> ```

Reviewer: Is this something you might want to do in middleware and/or a Server Action/Route Handler?

For the session management examples, it might be helpful to look at how `iron-session` handles it: https://get-iron-session.vercel.app/

They have examples using App Router.
